### PR TITLE
sum_positive_number_in_list

### LINKED ISSUE #29: sum_positive_number_in_list
# Sum the Positive Numbers in a List
Write a function that takes in a list of numbers and returns the sum of all positive numbers.

### PR DESCRIPTION
**Pull request by hudaalamassi**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_positive_numbers.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        >>> sum_positive_numbers([0, 0, 0])
+        0
+    """
+
+    # Validate input type
+    assert isinstance(numbers, list), "Input must be a list"
+
+    for num in numbers:
+        assert isinstance(num, (int, float)), (
+            f"List must contain only numeric elements, but found: {type(num)}"
```
Comment (mohamad-755) [inline solutions/sum_positive_numbers.py]: I think we can improve this because It’s better to raise explicit exceptions like TypeError or ValueError.
it can be improved as following 
```
if not isinstance(numbers, list):
    raise TypeError("Input must be a list")
for num in numbers:
    if not isinstance(num, (int, float)):
        raise TypeError(f"List must contain only numeric elements, but found: {type(num)}")
```

File: solutions/sum_positive_numbers.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        )
+
+    # Ensures that if there's no data to process, we handle it efficiently
+    if not numbers:
+        return 0
```
Comment (mohamad-755) [inline solutions/sum_positive_numbers.py]: The statement if not numbers: return 0 is unnecessary because an empty list ([]) naturally has a sum of 0. 
If you want you can combine validation and summation together so the list is iterated only once.

File: solutions/tests/test_sum_positive_numbers.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    # Test defensive assertions
+    def test_defensive_assertion_for_non_list_input(self):
+        """Test that an assertion is raised if the input is not a list."""
+        with self.assertRaises(AssertionError):
+            sum_positive_numbers(123)
+
+    def test_defensive_assertion_for_non_numeric_elements(self):
+        """Test that an assertion is raised if the list contains non-numeric elements."""
+        with self.assertRaises(AssertionError):
+            sum_positive_numbers([1, "a", 3])  # list contains a non-numeric element
```
Comment (mohamad-755) [inline solutions/tests/test_sum_positive_numbers.py]: dd a main block to allow running the tests directly

File: solutions/tests/test_sum_positive_numbers.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    # Test defensive assertions
+    def test_defensive_assertion_for_non_list_input(self):
+        """Test that an assertion is raised if the input is not a list."""
+        with self.assertRaises(AssertionError):
+            sum_positive_numbers(123)
+
+    def test_defensive_assertion_for_non_numeric_elements(self):
+        """Test that an assertion is raised if the list contains non-numeric elements."""
+        with self.assertRaises(AssertionError):
```
Comment (mohamad-755) [inline solutions/tests/test_sum_positive_numbers.py]: Replace AssertionError with TypeError in the tests to align with the suggested improvements in the function.

Comment (mohamad-755) [reply solutions/tests/test_sum_positive_numbers.py]: also include a test for an empty string or None as inputs to ensure defensive validation works as expected.

**Review by mohamad-755 — CHANGES_REQUESTED** [review]

**Review by Asia-Elfadil — COMMENTED** [review]
The code is clear and understandable, I have no comments to make. I will approve the merge

File: solutions/sum_positive_numbers.py (hunk 1 above)
Comment (hudaalamassi) [reply solutions/sum_positive_numbers.py]: - After researching the difference between assertions and explicit exceptions, I think assertions are primarily used to verify conditions assumed to be true during development.

- I also believe explicit exceptions, like TypeError or ValueError, can be more specific and valuable for error handling, as they provide clearer feedback on what went wrong. 

- However, assertions are important during debugging to quickly catch assumptions that might not hold. Since the rule specifically requires defensive assertions, I think I will stick to using assertions for checking assumptions about arguments.

- Thank you for pointing this, I think your opinion is valid, and explicit exceptions do offer more control and precision in error handling.

File: solutions/sum_positive_numbers.py (hunk 2 above)
Comment (hudaalamassi) [reply solutions/sum_positive_numbers.py]: - I think your point about removing if not numbers: return 0 is valid.
- I will remove it.

File: solutions/tests/test_sum_positive_numbers.py (hunk 3 above)
Comment (hudaalamassi) [reply solutions/tests/test_sum_positive_numbers.py]: - I'm not completely sure about the necessity of the main block, but after researching and considering our use of unittest and the focus on Test-Driven Development (TDD), I believe it might not be necessary in this case. 

- The main idea with TDD is to write code by ensuring that the tests pass as we develop. 

- Therefore, I don't think adding a main block is required here, but I appreciate your suggestion.

File: solutions/tests/test_sum_positive_numbers.py (hunk 4 above)
Comment (hudaalamassi) [reply solutions/tests/test_sum_positive_numbers.py]: - As I replied into the function file suggestion: since the rule specifically requires defensive assertions, I think I will stick to using assertions for checking assumptions about arguments.

- I will include another test as you suggested.

**Review by mohamad-755 — APPROVED** [review]
